XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,419 @@
+process quickly as to prevent an unintentional abort in the future. This is
+subject to change in the future if a comprehensive analysis is performed to
+ensure an exception can follow the normal control flow without unintentional
+side effects.
+
+
+#### `asyncHook.scope()`
+
+Enable capture of asynchronous events until the current synchronous code
+execution has completed. This is basically a small amount of sugar to prevent
+accidentally forgetting to `disable()` a set of hooks in a function with
+multiple returns. Also in case an error is thrown and caught by a domain or
+`uncaughtException`.
+
+Say an application only wishes to observe the asynchronous calls made from
+another asynchronous callback. Without `scope()` the call would look like so:
+
+```js
+fs.readFile(path, (err, data) => {
+  asyncHook.enable();
+  var e = asyncDataProcessing(data);
+  if (e) {
+    asyncHook.disable();
+    return;
+  }
+  doMoreAsyncStuff();
+  asyncHook.disable();
+});
+```
+
+But using `scope()` all of that is handled for you:
+
+```js
+fs.readFile(path, (err, data) => {
+  asyncHook.scope();
+  var e = asyncDataProcessing(data);
+  if (e)
+    return;
+  doMoreAsyncStuff();
+});
+```
+
+
+#### `asyncHook.enable()`
+
+Enable the callbacks for a given `AsyncHook` instance. Once a callback fires on
+an asynchronous event they will continue to fire for all nested asynchronous
+events. Even after the instance has been disabled.
+
+Callbacks are not implicitly enabled after an instance is created. The reason
+for this is to not make any assumptions about the user's use case. Since
+constructing the `asyncHook` during startup, but not using it until later is, a
+perfectly reasonable use case. This API is meant to err on the side of
+requiring explicit instructions from the user.
+
+
+#### `asyncHook.disable()`
+
+Disable the callbacks for a given `AsyncHook` instance. Doing this will prevent
+the `init()`, etc., calls from firing for any new roots of asynchronous call
+stacks, but will not prevent existing asynchronous call stacks that have
+already been captured by the `AsyncHook` instance from continuing to fire.
+
+While not part of the immediate development plan, it should be possible in the
+future to allow selective tracking of asynchronous call stacks. The following
+example demonstrates this:
+
+```js
+const async_wrap = require('async_wrap');
+const net = require('net');
+
+// Pretend init, pre, post are all defined
+const asyncHook = async_wrap.createHook({ init, pre, post });
+asyncHook.enable();
+
+net.createServer((c) => {
+  // Only want to follow connections that match IP range.
+  if (ipRangeRegExp.test(c.address().address))
+    asyncHook.disable();
+}).listen(PORT);
+```
+
+At which point no further calls to the hooks on that instance will be made for
+that asynchronous branch.
+
+
+### Hook Callbacks
+
+Key events in the lifetime of asynchronous events have been categorized into
```

Comment:
Given that you've defined handle/request above, I think it would be beneficial to describe the API in those terms consistently.  Most importantly is with the naming of init()/pre()/post()/destroy() be changed to indicate it's applicability on the handle/request.  E.g., names like initHandle()/destoryHandle()/preRequest()/postRequest() reinforce the mental model of the API, and IMO result in an  API that is more self-descriptive. My .02.  :)